specification/resources/resource-manager/Microsoft.Authorization/stable/2023-04-01/policyDefinitions.json

```diff
@@ -0,0 +1,732 @@
+        }
+      },
+      "description": "An error response from a policy operation."
+    },
+    "PolicyDefinitionProperties": {
+      "type": "object",
+      "properties": {
+        "policyType": {
+          "type": "string",
+          "description": "The type of policy definition. Possible values are NotSpecified, BuiltIn, Custom, and Static.",
+          "enum": [
+            "NotSpecified",
+            "BuiltIn",
+            "Custom",
+            "Static"
+          ],
+          "x-ms-enum": {
+            "name": "policyType",
+            "modelAsString": true
+          }
+        },
+        "mode": {
+          "type": "string",
+          "description": "The policy definition mode. Some examples are All, Indexed, Microsoft.KeyVault.Data.",
+          "default": "Indexed"
+        },
+        "displayName": {
+          "type": "string",
+          "description": "The display name of the policy definition."
+        },
+        "description": {
+          "type": "string",
+          "description": "The policy definition description."
+        },
+        "policyRule": {
+          "type": "object",
+          "description": "The policy rule."
+        },
+        "metadata": {
+          "type": "object",
+          "description": "The policy definition metadata.  Metadata is an open ended object and is typically a collection of key value pairs."
+        },
+        "parameters": {
+          "description": "The parameter definitions for parameters used in the policy rule. The keys are the parameter names.",
+          "$ref": "#/definitions/ParameterDefinitions"
+        },
+        "version": {
+          "type": "string",
+          "description": "The policy definition version in #.#.# format."
+        },
+        "versions": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "A list of available versions for this policy definition."
+        }
+      },
+      "description": "The policy definition properties."
+    },
+    "ParameterDefinitionsValue": {
+      "type": "object",
+      "properties": {
+        "type": {
+          "description": "The data type of the parameter.",
+          "type": "string",
+          "enum": [
+            "String",
+            "Array",
+            "Object",
+            "Boolean",
+            "Integer",
+            "Float",
+            "DateTime"
+          ],
+          "x-ms-enum": {
+            "name": "parameterType",
+            "modelAsString": true
+          }
+        },
+        "allowedValues": {
+          "type": "array",
+          "items": {
+            "type": "object"
+          },
+          "x-ms-identifiers": [],
+          "description": "The allowed values for the parameter."
+        },
+        "defaultValue": {
+          "type": "object",
+          "description": "The default value for the parameter if no value is provided."
+        },
+        "schema": {
```

Comment:
>schema [](http://example.com/codeflow?start=9&length=6)

ARM recommends against the use of the additionalProperties/empty/unschematized object pattern. ARM favors stronger typing wherever possible because it allows all consumers of your API to reason about it. This has benefits for everyone that interacts with your API including basic infrastructure, other Azure features, SDK owners, and customers. Unschematized objects are essentially API contracts that bypass Azure versioning infrastructure.

For example, customers will not be able to manage these objects using Azure Policy, or query them using ARG, both of which are standard Azure features. Customers expect to be able to use all common Azure features against all of their resources. 

Generally, this pattern should only be used when needed and the object properties are dynamic, unknown, or user defined. Ref: [https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined).

Are the values for this property user-defined?

Reply:
Yes these values are user defined.  This already exists and is/has been supported for some time, but for some reason was missed before.  This is just reflecting reality.